### DISCUSSION TITLE
Realm and ASP.NET Core

### DISCUSSION THREAD

**Discussion by darcome** [realm/realm-dotnet#3331] [Q&A] [answered]
Hi everyone,

I'd like to know if realm.net can be used as the backend for an ASP.NET Core website... 

1) Does it allow concurrency?
2) Are there any examples on how to setup it and use as a service?
3) Are there any pitfalls?
4) What would be its performance?
5) Anything else I should know? :)

Thanks in advance!

**Comment by nirinchev** [comment, marked as answer]
Hey, yes, Realm is supported on ASP.NET Core, even if it's not our primary use case.

1. Yes, concurrency is supported, though you need to be careful to always open a new Realm instance per thread (typically, that'll translate per request in ASP.NET).
2. There's nothing special about setting it up - you'll probably want to set it up to work with whatever DI framework you use to open and close a new Realm instance for each request.
3. The main pitfall is that a single Realm instance may only be used on the same synchronization context/thread. Since ASP.NET by default doesn't have a synchronization context installed, you'll either need to make sure you don't do any async work in the request handler or install a synchronization context. You can check out the ["Use Realm in a Console App" docs page](https://www.mongodb.com/docs/realm/sdk/dotnet/async-console/) as the information there is relevant for ASP.NET Core apps as well.
4. Performance should be as expected for an embedded database. It's quite fast, but if you're storing terabytes of data, you'll probably want to look into MongoDB or another server-side database.
5. Since Realm runs in the same process as your application, you may run into issues if you want to scale your web server across multiple physical servers. You'll need to set up your infrastructure in a way where they are able to access the same file, which may be tricky and non-performant due to the locking involved.

Generally speaking, Realm would be a good fit for single-server cache-like solution, but if you're load balancing a fleet of web servers, you're probably better-off with a dedicated database server.

**Reply by darcome** [reply]
Thank you very much for your answer!
